// @atproto/common@0.4.4/dist/ipld.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifyCidError = exports.VerifyCidTransform = exports.sha256RawToCid = exports.sha256ToCid = exports.verifyCidForBytes = exports.cborBytesToRecord = exports.isValidCid = exports.cidForCbor = exports.dataToCborBlock = exports.cborDecode = exports.cborEncode = void 0;
const crypto_1 = __importDefault(require("crypto"));
const stream_1 = require("stream");
const common_web_1 = require("@atproto/common-web");
const cid_1 = require("multiformats/cid");
const Block = __importStar(require("multiformats/block"));
const rawCodec = __importStar(require("multiformats/codecs/raw"));
const sha2_1 = require("multiformats/hashes/sha2");
const mf = __importStar(require("multiformats"));
const cborCodec = __importStar(require("@ipld/dag-cbor"));
exports.cborEncode = cborCodec.encode;
exports.cborDecode = cborCodec.decode;
const dataToCborBlock = async (data) => {
    return Block.encode({
        value: data,
        codec: cborCodec,
        hasher: sha2_1.sha256,
    });
};
exports.dataToCborBlock = dataToCborBlock;
const cidForCbor = async (data) => {
    const block = await (0, exports.dataToCborBlock)(data);
    return block.cid;
};
exports.cidForCbor = cidForCbor;
const isValidCid = async (cidStr) => {
    try {
        const parsed = cid_1.CID.parse(cidStr);
        return parsed.toString() === cidStr;
    }
    catch (err) {
        return false;
    }
};
exports.isValidCid = isValidCid;
const cborBytesToRecord = (bytes) => {
    const val = (0, exports.cborDecode)(bytes);
    if (!common_web_1.check.is(val, common_web_1.schema.map)) {
        throw new Error(`Expected object, got: ${val}`);
    }
    return val;
};
exports.cborBytesToRecord = cborBytesToRecord;
const verifyCidForBytes = async (cid, bytes) => {
    const digest = await sha2_1.sha256.digest(bytes);
    const expected = cid_1.CID.createV1(cid.code, digest);
    if (!cid.equals(expected)) {
        throw new Error(`Not a valid CID for bytes. Expected: ${expected.toString()} Got: ${cid.toString()}`);
    }
};
exports.verifyCidForBytes = verifyCidForBytes;
const sha256ToCid = (hash, codec) => {
    const digest = mf.digest.create(sha2_1.sha256.code, hash);
    return cid_1.CID.createV1(codec, digest);
};
exports.sha256ToCid = sha256ToCid;
const sha256RawToCid = (hash) => {
    return (0, exports.sha256ToCid)(hash, rawCodec.code);
};
exports.sha256RawToCid = sha256RawToCid;
class VerifyCidTransform extends stream_1.Transform {
    constructor(cid) {
        super();
        Object.defineProperty(this, "cid", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: cid
        });
        Object.defineProperty(this, "hasher", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: crypto_1.default.createHash('sha256')
        });
    }
    _transform(chunk, _enc, cb) {
        this.hasher.update(chunk);
        cb(null, chunk);
    }
    _flush(cb) {
        try {
            const cid = (0, exports.sha256RawToCid)(this.hasher.digest());
            if (this.cid.equals(cid)) {
                return cb();
            }
            else {
                return cb(new VerifyCidError(this.cid, cid));
            }
        }
        catch (_err) {
            const err = _err instanceof Error
                ? _err
                : new Error('Unexpected error', { cause: _err });
            return cb(err);
        }
    }
}
exports.VerifyCidTransform = VerifyCidTransform;
class VerifyCidError extends Error {
    constructor(expected, actual) {
        super('Bad cid check');
        Object.defineProperty(this, "expected", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: expected
        });
        Object.defineProperty(this, "actual", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: actual
        });
    }
}
exports.VerifyCidError = VerifyCidError;
